cpp/src/fil/infer.cu

```diff
@@ -248,20 +248,33 @@ struct tree_aggregator_t<NITEMS, GROVE_PER_CLASS_FEW_CLASSES> : finalize_block {
     __syncthreads();  // free up input row
     // load margin into shared memory
     auto per_thread = (vec<NITEMS, float>*)tmp_storage;
-    if (threadIdx.x >= num_classes) per_thread[threadIdx.x] = acc;
+    per_thread[threadIdx.x] = acc;
 
     __syncthreads();
     // reduce per-thread margin summand into per-class complete margin
     // (for each of the NITEMS rows)
-    // TODO(levsnv): use CUB/tree reduction when num_classes is small
-    if (threadIdx.x < num_classes) {
-      for (int c = threadIdx.x + num_classes; c < blockDim.x; c += num_classes)
-        acc += per_thread[c];
+    int t = threadIdx.x;
```

Comment:
If there is a separate function for such a reduction, would it be possible to have separate tests for it?

Reply:
currently, the coverage is provided by many different n_classes in fil_test.cu, which cover all %4 cases. Where would we test this function? Would it be fil_test.cu, as a completely separate test/kernel?

Reply:
Yes, that would be the best option. The test parameter structure would be separate as well. In a different pull request, the implementation and tests can be moved to primitives.